NXDOC-2956: document Audit Router and Blue/Green Audit migration (#2715)
nuxeo/doc.nuxeo.com-platform-spaces: src/nxdoc/nuxeo-server/data-store/data-modeling/audit.md

Author: I can't easily review this as you change the "format", for instance paragraph starting with `The Audit Service is used for logging and retrieving audit data into a data...` was previously a one-line, or maybe not but I can't see easily if you change something here.

On the other hand I like the new format you use to write the documentation. So keep it for new stuffs, but for old stuff I would rather you keeping the existing formatting. Could you do that so I can review?

Author: About explorer links, could you update them? New one looks like https://explorer.nuxeo.com/nuxeo/site/distribution/Nuxeo%20Platform-2025.8/viewExtensionPoint/org.nuxeo.audit.service.AuditComponent--event

Author: Could you update the code to use AuditService + getAuditBackend(DEFAULT_AUDIT_BACKEND)?

Author: Could you update docUUID and repositoryId to the appropriate constants from `LogEntryConstants`?

following implementations listed below:

| Backend                                  | Marketplace Package             | Reference                                                                          |
|------------------------------------------|---------------------------------|------------------------------------------------------------------------------------|
| In-Memory (not for production)           | Built-in Nuxeo Server (default) | —                                                                                  |
| OpenSearch 1.x / Elasticsearch 7.x - 8.x | nuxeo-audit-opensearch1         | [OpenSearch 1.x Audit Backend]({{page page='audit-backend-opensearch1'}})          |
| OpenSearch 2.x                           | nuxeo-audit-opensearch2         | [OpenSearch 2.x Audit Backend]({{page page='audit-backend-opensearch2'}})          |
| Elasticsearch 9.x                        | nuxeo-audit-elasticsearch9      | [Elasticsearch 9.x Audit Backend]({{page page='audit-backend-elasticsearch9'}})    |
| MongoDB                                  | nuxeo-audit-mongodb             | [MongoDB Audit Backend]({{page page='audit-backend-mongodb'}})                     |

> | SQL DataBase (Legacy)                    | nuxeo-audit-sql                 | [Legacy SQL Audit Backend]({{page page='audit-backend-sql'}})                      |

Reviewer: ```suggestion
| SQL Database (Legacy)                    | nuxeo-audit-sql                 | [SQL Audit Backend (Legacy)]({{page page='audit-backend-sql'}})                      |
```
First suggestion is about the existing, second one for consistency with the actual page title.

Each child page lists the configuration properties exposed by the package as
`nuxeo.audit.backend.default.<implementation>.*` so you can override them in
`nuxeo.conf`.

### Audit Router

The [Audit Router]({{page page='audit-router'}}) (since 2025.16) replaces the
legacy `event` extension point with a richer `routes` extension point:

* A route binds a **backend** to a set of **events** and **predicates**.
* Several routes can match the same entry, so the same event can be written to multiple backends in parallel.
* The legacy `event` extension point still works and is automatically merged into the implicit `default` route.

The [Audit Router]({{page page='audit-router'}}) child page contains the full
extension-point reference and worked examples (route a custom business event to
a secondary backend, filter by category, etc.).

## Querying the Audit Data Store

> The `AuditService` is the entry point to read audit data. From it you get an

Reviewer: read and write?

Author: Not exactly, there's currently no public write action API on `AuditService`. Its recent introduction in 2025 was focused on read API and still currently the `AuditService` doesn't have an API to write an `LogEntry` to the audit. The `AuditRouter` now offers a better way to insert `LogEntry` to the audit but it is more an internal service.

We will have one, one day, and the introduced `AuditRouter` will play a role there. This doesn't exist yet because we don't need it in our codebase. But still useful to offer for Customer project that does it at `AuditBackend` level, it works but presents limitations.

// Same method but with a query builder
AuditQueryBuilder builder = new AuditQueryBuilder();
builder.predicates(Predicates.eq(LogEntryConstants.LOG_DOC_UUID, doc.getId()),
                   Predicates.eq(LogEntryConstants.LOG_REPOSITORY_ID, "myRepository"));
List<LogEntry> logEntriesFiltered = backend.queryLogs(builder);
```

To target another backend, just ask the `AuditService` for it:

```java
AuditBackend secondary = auditService.getAuditBackend("secondary");
List<LogEntry> entries = secondary.queryLogs(builder);
```

There are two PageProviders that can be used for querying the Audit data store:

* `AuditPageProvider`: allows to generate simple queries against Audit entries.
* `DocumentHistoryReader`: allows to compute history for a given document.

    [More details on the explorer](http://explorer.nuxeo.com/nuxeo/site/latest/viewContribution/org.nuxeo.audit.PageProviderService.contrib--providers).

By default both PageProviders query the default backend. They support targeting
a specific backend through the `backend` PageProvider property.

A schema has been defined for basic Audit search: `basicauditsearch.xsd`. This

> schema is helpful for building a PageProvider feeding a ContentView with data

Reviewer: Worth renaming "ContentView", which is a JSF remainder

| `extendedInfo`   | Extract custom values from the event context using EL and store them on the log entry.                                  |
| `adapter`        | Register an adapter exposed to EL expressions of `extendedInfo`.                                                        |
| `backendFactory` | Register an `AuditBackend` implementation. Contributed by the audit Marketplace Packages, rarely used by integrations.  |

### Routes — Recommended Since 2025.16

The `routes` extension point is the modern way to declare what is audited and
where it is stored. A minimal route registers an event and writes it to the
default backend:

```xml
<extension target="org.nuxeo.audit.service.AuditComponent" point="routes">
  <route name="documentLifecycle">
    <backend name="default" />
    <event name="documentCreated" />
    <event name="documentModified" />
  </route>
</extension>
```

See the [Audit Router]({{page page='audit-router'}}) child page for the full
reference (predicates, route inheritance with `copy`, routing semantics, …) and
worked examples.

### Event — Deprecated Since 2025.16

The `event` extension point keeps registering auditable events on the implicit

> `default` route. It is the historical way of declaring audited events and

Reviewer: ```suggestion
`default` route. It is the legacy way of declaring audited events and
```
?

## Architecture

The Audit Service is mainly a data store service. It defines a data record
structure that will be used for storing audit information.

The data record structure is defined in Java by the `LogEntry` and
`ExtendedInfo` Java classes. The Audit Service receives events from the Event
Service. Then the Audit Service filters and converts them into log entries. The
`LogEntry` class is mainly obtained from a `DocumentEventContext`.

Nuxeo documents and events can have a lot of custom properties, so if you want
to log some specific events or document properties, the
[Extended Info](#extendedinfo) allows for a Key/Value type storage that will be

> associated to the main `LogEntry` record. These informations are extracted from

Reviewer 2: ```suggestion
associated to the main `LogEntry` record. This information is extracted from
```

> the event message using and EL (Expression Language) expression and stored into

Reviewer 3: ```suggestion
the event message using an EL (Expression Language) expression and stored into
```

the event message using and EL (Expression Language) expression and stored into
a map.

Since LTS 2025 the Audit Service has been redesigned to support **multiple
backends in parallel** through the
[`AuditRouter`]({{page page='audit-router'}}). The high-level pipeline is:

```
Nuxeo Event Service
        │
        ▼
StreamAuditEventListener (sync, priority 500)
        │  AuditRouter computes LogEntry from event
        ▼
audit/audit Nuxeo Stream
        │
        ▼
audit/writer computation (StreamAuditWriter)
        │  AuditRouter.routeToBackend(LogEntry)
        ▼
        ├──► Audit Backend "default"
        ├──► Audit Backend "secondary"
        └──► …
```

### Services

| Service                                              | Purpose                                                          |
|------------------------------------------------------|------------------------------------------------------------------|
| `org.nuxeo.audit.service.AuditService`               | Entry point — exposes the configured backends and the router.     |
| `org.nuxeo.audit.service.AuditBackend`               | Backend SPI — read, write and query log entries.                  |

> | `org.nuxeo.audit.service.AuditRouter`                | Computes the list of target backends for a given log entry.       |

Author: Purpose could be rephrased like internal service to compute and route log entries.

The legacy `AuditReader`, `AuditLogger` and `AuditAdmin` services from the
`org.nuxeo.ecm.platform.audit.api` package are still available and proxy to the
default backend. They remain the simplest way to read entries from a single

> backend; new code should target `AuditService` and `AuditBackend` directly.

Author: They don't remain the simplest way, those will be removed in a future version, talk about that and maybe print a bit of code to retrieve the default AuditBackend.

| MongoDB                                  | nuxeo-audit-mongodb             | [MongoDB Audit Backend]({{page page='audit-backend-mongodb'}})                     |
| SQL DataBase (Legacy)                    | nuxeo-audit-sql                 | [Legacy SQL Audit Backend]({{page page='audit-backend-sql'}})                      |

Each child page lists the configuration properties exposed by the package as
`nuxeo.audit.backend.default.<implementation>.*` so you can override them in
`nuxeo.conf`.

### Audit Router

The [Audit Router]({{page page='audit-router'}}) (since 2025.16) replaces the
legacy `event` extension point with a richer `routes` extension point:

* A route binds a **backend** to a set of **events** and **predicates**.
* Several routes can match the same entry, so the same event can be written to multiple backends in parallel.
* The legacy `event` extension point still works and is automatically merged into the implicit `default` route.

The [Audit Router]({{page page='audit-router'}}) child page contains the full
extension-point reference and worked examples (route a custom business event to
a secondary backend, filter by category, etc.).

> ### Blue/Green Migration — Copy an Audit Backend

Author: I'm not sure it needs to be stated in this page, please remove this block.

Since 2025.19 Nuxeo provides building blocks to copy log entries from one
backend to another without downtime:

* An **`audit` scroll** based on the Audit `QueryBuilder` API.
* A **`copyAudit` Bulk Action** that reads the scroll and writes every entry to a target backend.
* Three Management REST endpoints ([Audit Endpoint]({{page space='rest-api' version='1' page='audit-endpoint'}})) to trigger a copy, compare two backends and introspect the live router topology.

The `GET /management/audit/introspection` endpoint returns a PlantUML
representation of the live audit routing — useful to inspect the routes
contributed by all the deployed bundles, regardless of how they were registered.

See [Copy an Audit Backend]({{page page='copy-audit-backend'}}) for details and
the typical Blue/Green migration workflow.

> ## Querying the Audit Data Store

Author: Please review this block, it states services that are deprecated.

List<LogEntry> logEntriesFiltered = reader.queryLogs(builder);
```

To target a specific backend, get an `AuditBackend` from the `AuditService`:

```java
AuditService auditService = Framework.getService(AuditService.class);
AuditBackend secondary = auditService.getAuditBackend("secondary");
List<LogEntry> entries = secondary.queryLogs(builder);

> ```

Author: This are great, as I see PageProvider section just after, maybe we need to add few words to address https://hyland.atlassian.net/browse/NXP-33467?

|------------------|------------------------------------------------------------------------------------------------------------------------|
| `routes`         | Bind a backend to a set of events and predicates. **Since 2025.16.**                                                   |
| `event`          | Register an auditable event. **Deprecated since 2025.16** — use `routes` instead. Backward-compatible.                  |
| `extendedInfo`   | Extract custom values from the event context using EL and store them on the log entry.                                  |
| `adapter`        | Register an adapter exposed to EL expressions of `extendedInfo`.                                                        |
| `backendFactory` | Register an `AuditBackend` implementation. Contributed by the audit Marketplace Packages, rarely used by integrations.  |

A separate extension point on
`org.nuxeo.ecm.core.persistence.PersistenceComponent` named `hibernate` is
documented on the
[Legacy SQL Audit Backend]({{page page='audit-backend-sql'}}#hibernate-configuration)

> page.

Author: Please remove this part.

> * `documentRestored`

Reviewer: I wonder if these shouldn't be in the Routes extension point section, since the actual way of declaring these auditbale events in 2025 is under the default route contribution?

Author: Yes it makes more sense, I will move & update the section below

```
Those default auditable events match the Nuxeo core base events:

* `documentCreated`
* `documentCreatedByCopy`
* `documentDuplicated`
* `documentMoved`
* `documentRemoved`
* `documentModified`
* `documentLocked`
* `documentUnlocked`
* `documentSecurityUpdated`
* `lifecycle_transition_event`
* `loginSuccess`
* `loginFailed`
* `logout`
* `documentCheckedIn`
* `versionRemoved`
* `documentProxyPublished`
* `sectionContentPublished`
* `documentRestored`

{{#> callout type='info'  heading='Full List'}}
The full list of audit events used in the platform can be found
[here](https://explorer.nuxeo.com/nuxeo/site/latest/viewExtensionPoint/org.nuxeo.audit.service.AuditComponent--event).
{{/callout}}

If you are sending new Nuxeo core events and want them to be audited, you have
to extend the `event` extension point. Here is an example of a contribution to
this extension point:

```xml
<extension target="org.nuxeo.audit.service.AuditComponent" point="event">
    <event name="myCustomEvent" />
</extension>
```

[More details on the explorer.](https://explorer.nuxeo.com/nuxeo/site/latest/viewExtensionPoint/org.nuxeo.audit.service.AuditComponent--event)
```

To _### Routes — Recommended Since 2025.16_ paragraph.

{{#> callout type='info'  heading='Full List'}}
The full list of audit events used in the platform can be found

> [here](https://explorer.nuxeo.com/nuxeo/site/latest/viewExtensionPoint/org.nuxeo.audit.service.AuditComponent--event).

Reviewer: Isn't it the "routes" extension point?

Reviewer: Also, it seems to be distribution/latest

Author: Yes, it would be according to the previous comment.

{{/callout}}

If you are sending new Nuxeo core events and want them to be audited, you have
to extend the `event` extension point. Here is an example of a contribution to
this extension point:

```xml
<extension target="org.nuxeo.audit.service.AuditComponent" point="event">
    <event name="myCustomEvent" />
</extension>
```

> [More details on the explorer.](https://explorer.nuxeo.com/nuxeo/site/latest/viewExtensionPoint/org.nuxeo.audit.service.AuditComponent--event)

Reviewer: ```suggestion
[More details on the explorer.](https://explorer.nuxeo.com/nuxeo/site/distribution/latest/viewExtensionPoint/org.nuxeo.audit.service.AuditComponent--event)
```

extension point

```xml

> <extension point="extendedInfo" target="org.nuxeo.audit.service.AuditComponent">

Author: Could you review the extension point of pages we worked on in order to have target then point?

</extension>
```

For instance, the above contribution will add `modelId`, `modelName`,
`worklowInitiator`, `workflowVarriables` to the `extendedInfo` only for the
`afterWorkflowStarted` event.

{{#> callout type='info' heading='Recommended since 2025.16'}}
Since the introduction of the [Audit Router]({{page page='audit-router'}}),

> per-event `extendedInfo` entries has to be declared directly on the

Reviewer: ```suggestion
per-event `extendedInfo` entries have to be declared directly on the
```

`extendedInfo` extension point with an `event` attribute, for instance:

```xml
<extension target="org.nuxeo.audit.service.AuditComponent" point="extendedInfo">
    <extendedInfo event="afterWorkflowStarted" expression="${message.properties.modelId}" key="modelId" />
</extension>
```

This is the recommended form. The legacy form nesting `<extendedInfo>` inside
`<event>` is kept for backward compatibility but will be removed in the future.
{{/callout}}

> [More details on the explorer.](https://explorer.nuxeo.com/nuxeo/site/latest/viewExtensionPoint/org.nuxeo.audit.service.AuditComponent--extendedInfo)

Reviewer: ```suggestion
[More details on the explorer.](https://explorer.nuxeo.com/nuxeo/site/distribution/latest/viewExtensionPoint/org.nuxeo.audit.service.AuditComponent--extendedInfo)
```

### Adapter

The contribution to the `adapter` extension point of the
`org.nuxeo.audit.service.AuditComponent` component allows to define the adapter
that will be injected in the EL context. Here is an example of a contribution to
this extension point.

```xml
<extension target="org.nuxeo.audit.service.AuditComponent" point="adapter">
    <adapter name="myadapter" class="org.nuxeo.ecm.core.api.facet.VersioningDocument"/>
</extension>
```

> [More details on the explorer.](https://explorer.nuxeo.com/nuxeo/site/latest/viewExtensionPoint/org.nuxeo.audit.service.AuditComponent--adapter)

Reviewer: ```suggestion
[More details on the explorer.](https://explorer.nuxeo.com/nuxeo/site/distribution/latest/viewExtensionPoint/org.nuxeo.audit.service.AuditComponent--adapter)
```

</extension>
```

For instance, the above contribution will add `modelId`, `modelName`,

> `worklowInitiator`, `workflowVarriables` to the `extendedInfo` only for the

Reviewer 2: ```suggestion
`workflowInitiator`, `workflowVariables` to the `extendedInfo` only for the
```

You can also extend the audit info per event name:

```xml

> <extension target="org.nuxeo.audit.service.AuditComponent" point="event">

Author: extendedInfo contribution per event should also be updated to reflect https://hyland.atlassian.net/browse/NXP-33469 as stated in the paragraph _Future Design / Contribution_ of https://hyland.atlassian.net/wiki/spaces/NuxEng/pages/3837330155/Architecture+Multiple+Audit+Backends

`worklowInitiator`, `workflowVarriables` to the `extendedInfo` only for the
`afterWorkflowStarted` event.

The way the data is persisted depends on the backend (column, JSON document,
mapping, …) — refer to the [backend pages](#audit-back-ends).

[More details on the explorer.](http://explorer.nuxeo.com/nuxeo/site/distribution/latest/viewExtensionPoint/org.nuxeo.ecm.platform.audit.service.NXAuditEventsService--extendedInfo)

### Adapter

The contribution to the `adapter` extension point of the
`org.nuxeo.audit.service.AuditComponent` component allows to define the adapter
that will be injected in the EL context. Here is an example of a contribution to
this extension point.

```xml
<extension target="org.nuxeo.audit.service.AuditComponent" point="adapter">
    <adapter name="myadapter" class="org.nuxeo.ecm.core.api.facet.VersioningDocument"/>
</extension>
```

[More details on the explorer.](http://explorer.nuxeo.com/nuxeo/site/distribution/latest/viewExtensionPoint/org.nuxeo.ecm.platform.audit.service.NXAuditEventsService--adapter)

### Live Ingestion — Stream Audit Writer

Since 2025 the audit ingestion pipeline is built on top of

> [Nuxeo Stream]({{page page='nuxeo-stream'}}). A synchronous event listener

Author: Not sure about the since 2025, just state the information without the since.
